src/Altinn.App.Core/Helpers/Serialization/ModelSerializationService.cs

```diff
@@ -0,0 +1,267 @@
+    /// </summary>
+    /// <param name="data">The binary data</param>
+    /// <param name="dataType">The data type used to get content type and the classRef for the object to be returned</param>
+    /// <returns>The model specified in </returns>
+    public object DeserializeFromStorage(ReadOnlySpan<byte> data, DataType dataType)
+    {
+        var type = GetModelTypeForDataType(dataType);
+
+        // TODO: support sending json to storage based on dataType.ContentTypes
+        return DeserializeXml(data, type);
+    }
+
+    /// <summary>
+    /// Serialize an object to binary data for storage, respecting classRef and content type in dataType
+    /// </summary>
+    /// <param name="model">The object to serialize (must match the classRef in DataType)</param>
+    /// <param name="dataType">The data type</param>
+    /// <returns>the binary data and the content type (currently only application/xml, but likely also json in the future)</returns>
+    /// <exception cref="InvalidOperationException">If the classRef in dataType does not match type of the model</exception>
+    public (ReadOnlyMemory<byte> data, string contentType) SerializeToStorage(object model, DataType dataType)
+    {
+        var type = GetModelTypeForDataType(dataType);
+        if (type != model.GetType())
+        {
+            throw new InvalidOperationException(
+                $"DataType {dataType.Id} expects {type.FullName}, found {model.GetType().FullName}"
+            );
+        }
+
+        //TODO: support sending json to storage based on dataType.ContentTypes
+        return (SerializeToXml(model), "application/xml");
+    }
+
+    /// <summary>
+    /// Serialize an object to xml
+    /// </summary>
+    /// <param name="model">The object to serialize</param>
+    /// <returns>The bytes of the serialized xml in UTF8 encoding</returns>
+    public ReadOnlyMemory<byte> SerializeToXml(object model)
+    {
+        var modelType = model.GetType();
+        using var activity = _telemetry?.StartSerializeToXmlActivity(modelType);
+
+        // Ensure that model is mutated in the same way it would be when deserialized from storage
+        // (evaluate ShouldSerialize* methods, set empty strings to null, etc.)
+        ObjectUtils.PrepareModelForXmlStorage(model);
+
+        XmlWriterSettings xmlWriterSettings = new XmlWriterSettings()
+        {
+            Encoding = new UTF8Encoding(false),
+            NewLineHandling = NewLineHandling.None,
+        };
+        using var memoryStream = new MemoryStream();
+        XmlWriter xmlWriter = XmlWriter.Create(memoryStream, xmlWriterSettings);
+
+        XmlSerializer serializer = _xmlSerializer.GetSerializer(modelType);
+        serializer.Serialize(xmlWriter, model);
+        if (!memoryStream.TryGetBuffer(out var segment))
+        {
+            throw new InvalidOperationException("Failed to get buffer from memory stream");
+        }
+
+        return segment.AsMemory().RemoveBom();
```

Comment:
I kind of want to avoid the need for copies, and it does not look like `MemoryStream` is renting buffers that gets recycled in dispose, but this should probably be simplified to
```suggestion
        return memoryStream.ToArray().AsMemory().RemoveBom();
```